.github/workflows/docker-build-test.yaml

```diff
@@ -307,17 +301,116 @@ jobs:
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0  # Fetch all history for all branches and tags
+
+      # To test the compat test against the last released docker image, we need to get the last released docker image tag.
+      # This step determines appropriate branch to fetch the latest docker image tag from based on the current branch.
+      # Determine the IMAGE_TAG based on the latest release R and check for images built on branch R - 1:
+      # - ex. For commits on the aptos-release-v1.19 branch, the IMAGE_TAG should be the latest commit built on aptos-release-v1.18
+      # - For commits on the main branch, the IMAGE_TAG should be the latest commit on the max release aptos-release-v<MAX> branch
+      - name: Determine branch to fetch latest docker image tag from
+        id: determine-the-target-branch
+        run: |
+          # Determine the base branch (the branch we're merging into)
+          if [[ "${{ github.event_name }}" == "pull_request" || "${{ github.event_name }}" == "pull_request_target" ]]; then
+            base_branch="${{ github.base_ref }}"
+          else
+            base_branch="${{ github.ref_name }}"
+          fi
+
+          if [[ $base_branch == "main" ]]; then
+            target_branch=$(git ls-remote --heads origin 'aptos-release-v*' | 
+                            grep -E 'refs/heads/aptos-release-v[0-9]+\.[0-9]+$' | 
+                            sed 's/.*refs\/heads\///' | 
+                            sort -t. -k2,2n -k3,3n | 
+                            tail -n 1)
+          elif [[ $base_branch =~ ^aptos-release-v([0-9]+)\.([0-9]+) ]]; then
+            major=${BASH_REMATCH[1]}
+            minor=${BASH_REMATCH[2]}
+
+            if [[ $minor -eq 0 ]]; then
+                # If the minor version is 0, look for the latest branch in the previous major version
+                prev_major=$((major - 1))
+                target_branch=$(git ls-remote --heads origin "aptos-release-v${prev_major}.*" | 
+                                grep -E "refs/heads/aptos-release-v${prev_major}\.[0-9]+$" | 
+                                sed 's/.*refs\/heads\///' | 
+                                sort -t. -k2,2n -k3,3n | 
+                                tail -n 1)
+            else
+                # If the minor version is greater than 0, look for the previous minor version
+                prev_minor=$((minor - 1))
+                target_branch="aptos-release-v${major}.${prev_minor}"
+            fi
+
+            # Now do the check whether the target branch exists, outside the if-else
+            if git ls-remote --heads origin "aptos-release-v${target_branch}" | grep -q "refs/heads/${target_branch}"; then
+                echo "Target branch is $target_branch"
+            else
+                echo "Branch $target_branch does not exist"
+            fi
+          else
+            echo "Not merging to main or a release branch, using the main branch for compat test" >&2
```

Comment:
@perryjrandall I think your case is still a "merge to main" case though. Since commit `c` was presumably created from a PR targeting `main`. 

Let's consider another structure to be a bit more clear

```
a
|    b1 -> aptos-release-banana (tag)
b_/
|    d2 (PR)
c   |
|    d1 -> d-personal-branch (branch)
d_/
|    e1 -> aptos-release-egret (tag)
e_/
```

`d1` is the tip of a random branch `d-personal-branch` that's branched off of `main` at some point, and we want to merge `d2` in a PR into `d-personal-branch` (on top of `d1`). This PR does not target `main` branch -- it targets `d-personal-branch`. In this case we should test the PR with `aptos-release-egret`, which is the latest release in the PR source branch's history

So basically for both PRs that target non-main/non-release branches and PRs that target main, they both test against the same thing: the latest release in the target branch's git history


We should probably do something like this:
* Try to find the latest release in the PR target branch's git history
* If we can't find it in like the last ~100s commits, we may want to emit an error (probably prompt the user to rebase since it's likely they forgot to do this). This assumes that there are no large working branches of that size.
Lmk if this does/doesn't make sense, happy to sync separately